linked_list/linked_list.py

```diff
@@ -9,71 +9,150 @@ def __init__(self, value, next_node = None):
+        
+        return length
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        if self.head is None:
+            return None
+        
+        current_node = self.head
+        i = 0
+        while i < index and current_node:
+            current_node = current_node.next
+            i+=1
+        
+        return current_node.value
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        if self.head is None:
+            return None
+        
+        current = self.head
+
+        while current:
+            if current.next is None:
+                return current.value
+            else:
+                current = current.next
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        new_node = Node(value)
+
+        if self.head is None:
+            self.head = new_node
+            return 
+
+        last_node = self.head
+
+        while last_node.next:
+            last_node = last_node.next 
+        
+        last_node.next = new_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_max(self):
-        pass
+        if self.head is None:
+            return None
+
+        current = self.head
+        max = self.head.value
+
+        while current.next:
+            if current.next.value > max:
+                max = current.next.value
+                current = current.next
+            else:
+                current = current.next
+
+        return max
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        if self.head is None:
+            return None
+
+        current = self.head
+        if current.value == value:
+            self.head = current.next
+            return
+
+        while current: 
+            if current.value == value:
+                current = current.next
+                return
+
+            prev = current
+            current = current.next
+            if current.value == value:
+                prev.next = current.next
+                return

```

Comment:
This looks right, but I think the prior `if` statement could cause an occasional subtle bug.